src/cmap/wire_protocol/on_demand/document.ts

```diff
@@ -0,0 +1,320 @@
+import {
+  Binary,
+  BSON,
+  type BSONElement,
+  BSONError,
+  type BSONSerializeOptions,
+  BSONType,
+  getBigInt64LE,
+  getFloat64LE,
+  getInt32LE,
+  ObjectId,
+  parseToElementsToArray,
+  Timestamp,
+  toUTF8
+} from '../../../bson';
+
+// eslint-disable-next-line no-restricted-syntax
+const enum BSONElementOffset {
+  type = 0,
+  nameOffset = 1,
+  nameLength = 2,
+  offset = 3,
+  length = 4
+}
+
+export type JSTypeOf = {
+  [BSONType.null]: null;
+  [BSONType.undefined]: null;
+  [BSONType.double]: number;
+  [BSONType.int]: number;
+  [BSONType.long]: bigint;
+  [BSONType.timestamp]: Timestamp;
+  [BSONType.binData]: Binary;
+  [BSONType.bool]: boolean;
+  [BSONType.objectId]: ObjectId;
+  [BSONType.string]: string;
+  [BSONType.date]: Date;
+  [BSONType.object]: OnDemandDocument;
+  [BSONType.array]: OnDemandDocument;
+};
+
+/** @internal */
+export class OnDemandDocument {
+  /** Caches the existence of a property */
+  private readonly existenceOf: Record<string, boolean> = Object.create(null);
+  /** Caches a look up of name to element */
+  private readonly elementOf: Record<string, BSONElement> = Object.create(null);
+  /** Caches the revived javascript value */
+  private readonly valueOf: Record<string, any> = Object.create(null);
+  /** Caches the index of elements that have been named */
+  private readonly indexFound: Record<number, boolean> = Object.create(null);
+
+  /** All bson elements in this document */
+  private readonly elements: BSONElement[];
+
+  /** The number of elements in the BSON document */
+  public readonly length: number;
```

Comment:
`numberOfElements`?

Reply:
I'd still lean toward `size()` here, but less is more, we can revist this if it is needed. For example, cursor might want to know how many docs are left but we should solve for that case.